File: benchmarks/fixtures/webapp_rb/database/pool.rb
# Database connection pool.

require_relative '../utils/helpers'
require_relative '../exceptions'

# Connection handle struct.
class ConnectionHandle
  attr_accessor :id, :created_at, :last_used, :in_use, :query_count

  def initialize(id:)
    @id = id
    @created_at = Time.now
    @last_used = Time.now
    @in_use = false
    @query_count = 0
  end
end

# Manages a pool of database connections.
class ConnectionPool
  def initialize(dsn, pool_size: 10)
    @dsn = dsn
    @pool_size = [pool_size, 50].min
    @connections = []
    @initialized = false
    get_logger('database.pool').info("Pool created: size=#{@pool_size}")
  end

  # Initialize the pool with connections.
  def do_initialize
    return if @initialized

    @pool_size.times do |i|
      @connections << ConnectionHandle.new(id: "conn-#{i}")
    end
    @initialized = true
    get_logger('database.pool').info("Pool initialized with #{@pool_size} connections")
  end

  # Acquire a connection from the pool.
  def get_connection
    do_initialize unless @initialized
    @connections.each do |conn|
      unless conn.in_use
        conn.in_use = true
        conn.last_used = Time.now
        conn.query_count += 1
        get_logger('database.pool').info("Acquired connection #{conn.id}")
        return conn
      end
    end
    raise DatabaseError.new('Connection pool exhausted')
  end

  # Release a connection back to the pool.
  def release_connection(handle)
    handle.in_use = false
    handle.last_used = Time.now
    get_logger('database.pool').info("Released connection #{handle.id}")
  end

  # Get pool statistics.
  def stats
    active = @connections.count(&:in_use)
    { total: @connections.length, active: active, idle: @connections.length - active }
  end

  # Shut down the pool.
  def shutdown
    @connections = []
    @initialized = false
    get_logger('database.pool').info('Pool shut down')
  end
end

File: benchmarks/fixtures/webapp_ts/src/database/pool.ts
/** Database connection pool. */
import { getLogger } from '../utils/helpers';
import { DatabaseError } from '../errors';

const logger = getLogger("database.pool");

export interface ConnectionHandle {
    id: string;
    createdAt: number;
    lastUsed: number;
    inUse: boolean;
    queryCount: number;
}

/** Manages a pool of database connections. */
export class ConnectionPool {
    private connections: ConnectionHandle[] = [];
    private poolSize: number;
    private initialized: boolean = false;

    constructor(private dsn: string, poolSize: number = 10) {
        this.poolSize = Math.min(poolSize, 50);
        logger.info(`Pool created: size=${this.poolSize}`);
    }

    /** Initialize the pool with connections. */
    initialize(): void {
        if (this.initialized) return;
        for (let i = 0; i < this.poolSize; i++) {
            this.connections.push({
                id: `conn-${i}`,
                createdAt: Date.now(),
                lastUsed: Date.now(),
                inUse: false,
                queryCount: 0,
            });
        }
        this.initialized = true;
        logger.info(`Pool initialized with ${this.poolSize} connections`);
    }

    /** Acquire a connection from the pool. */
    getConnection(): ConnectionHandle {
        if (!this.initialized) this.initialize();
        for (const conn of this.connections) {
            if (!conn.inUse) {
                conn.inUse = true;
                conn.lastUsed = Date.now();
                conn.queryCount++;
                logger.info(`Acquired connection ${conn.id}`);
                return conn;
            }
        }
        throw new DatabaseError("Connection pool exhausted");
    }

    /** Release a connection back to the pool. */
    releaseConnection(handle: ConnectionHandle): void {
        handle.inUse = false;
        handle.lastUsed = Date.now();
        logger.info(`Released connection ${handle.id}`);
    }

    /** Get pool statistics. */
    stats(): { total: number; active: number; idle: number } {
        const active = this.connections.filter(c => c.inUse).length;
        return { total: this.connections.length, active, idle: this.connections.length - active };
    }

    /** Shut down the pool. */
    shutdown(): void {
        this.connections = [];
        this.initialized = false;
        logger.info("Pool shut down");
    }
}
